applications/callflow/src/module/cf_voicemail.erl

```diff
@@ -1408,26 +1414,30 @@ change_pin(#mailbox{mailbox_id=Id
                 Box;
             {'error', _Reason} ->
                 lager:debug("box failed validation: ~p", [_Reason]),
-                invalid_pin(Box, Call)
+                invalid_pin(Box, Call, Loop)
         end
     catch
         'error':{'badmatch',{'error','channel_hungup'}} ->
             lager:debug("channel hungup while configuring pin"),
             {'error', 'channel_hungup'};
         'error':{'badmatch',{'ok',_ConfirmPin}} ->
             lager:debug("new pin was invalid, try again"),
-            invalid_pin(Box, Call);
+            invalid_pin(Box, Call, Loop);
         _E:_R ->
             lager:debug("failed to get new pin: ~s: ~p", [_E, _R]),
-            invalid_pin(Box, Call)
+            invalid_pin(Box, Call, Loop)
     end.
 
--spec invalid_pin(mailbox(), kapps_call:call()) ->
+-spec invalid_pin(mailbox(), kapps_call:call(), non_neg_integer()) ->
                          mailbox() |
                          {'error', any()}.
-invalid_pin(Box, Call) ->
+invalid_pin(_Box, Call, Loop) when Loop > 2 -> %% third failure, terminate
+    lager:debug("channel hungup after several empty or invalid pins"),
+    _ = kapps_call_command:b_prompt(<<"vm-goodbye">>, Call),
+    {'error', 'channel_hungup'};
```

Comment:
Isn't hanging up here too harsh? You can just return `Box` here and it goes back to config menu.
Also instead of playing `vm-goodbye`, play `vm-pin_invalid`.

Reply:
@lazedo Config menu has it's own loop and will stop.
I've notice config menu is going to crash if it passes 3 loop. Shouldn't it just return to main menu(and again menu has its own loop)?

Reply:
@icehess whatif its not from config menu ?

Reply:
@lazedo invalid_pin is only called by change_pin and config_menu

Reply:
@harenson So you can play `vm-pin_invalid` here and return `{'error', 'max_retry'}`.
And in `setup_mailbox/2` check `change_pin/2` result if it is an error play `vm-goodbye` and return `{'error', 'channel_hungup'}`.

@lazedo What do you think?